Ansible role and ansible playbook to uninstall openebs using helm chart.

gprasath opened the pull request:
Signed-off-by: gprasath <giridhara.prasad@cloudbyte.com>

Implemented ansible role to uninstall openebs using helm chart and including this role in the cleanup-openebs playbook.

**Steps followed:**

- Find out the openebs chart name.
-  Uninstall openebs.
-  Verify if openebs is uninstalled successfully.
-  Delete the namespace.
-  Uninstall helm.
-  Delete the resources service account and clusterrolebinding.
-  Delete the test artifacts.

Refer the issue: https://github.com/openebs/openebs/issues/1569

**SAMPLE TEST RUN OUTPUT:**

```
giri@vagrant-host:~/openebs/e2e/ansible$ ansible-playbook cleanup-openebs.yml -vv
ansible-playbook 2.4.3.0
  config file = /home/giri/openebs/e2e/ansible/ansible.cfg
  configured module search path = [u'/home/giri/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
Using /home/giri/openebs/e2e/ansible/ansible.cfg as config file

PLAYBOOK: cleanup-openebs.yml ******************************************************************************************************************************************
1 plays in cleanup-openebs.yml

PLAY [kubernetes-kubemasters] ******************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************************
ok: [kubemaster01]
META: ran handlers

TASK [k8s-openebs-uninstall-helm : Getting OpenEBS chart name] *********************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:3
changed: [kubemaster01] => {"changed": true, "cmd": "helm ls --all | grep openebs | grep DEPLOYED | awk '{print $1}'", "delta": "0:00:01.077105", "end": "2018-05-31 16:00:32.367665", "rc": 0, "start": "2018-05-31 16:00:31.290560", "stderr": "", "stderr_lines": [], "stdout": "openebs", "stdout_lines": ["openebs"]}

TASK [k8s-openebs-uninstall-helm : Uninstall OpenEBS] ******************************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:9
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "helm del --purge \"openebs\"", "delta": "0:00:11.338293", "end": "2018-05-31 16:00:44.606582", "rc": 0, "start": "2018-05-31 16:00:33.268289", "stderr": "", "stderr_lines": [], "stdout": "release \"openebs\" deleted", "stdout_lines": ["release \"openebs\" deleted"]}

TASK [k8s-openebs-uninstall-helm : Checking if provisioner and api server are deleted.] ********************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:18
FAILED - RETRYING: Checking if provisioner and api server are deleted. (15 retries left).
changed: [kubemaster01] => {"attempts": 2, "changed": true, "cmd": "kubectl get po -n openebs", "delta": "0:00:01.020544", "end": "2018-05-31 16:01:50.461093", "rc": 0, "start": "2018-05-31 16:01:49.440549", "stderr": "No resources found.", "stderr_lines": ["No resources found."], "stdout": "", "stdout_lines": []}

TASK [k8s-openebs-uninstall-helm : Deleting the namespace] *************************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:27
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "kubectl delete ns openebs", "delta": "0:00:00.567794", "end": "2018-05-31 16:01:52.155611", "rc": 0, "start": "2018-05-31 16:01:51.587817", "stderr": "", "stderr_lines": [], "stdout": "namespace \"openebs\" deleted", "stdout_lines": ["namespace \"openebs\" deleted"]}

TASK [k8s-openebs-uninstall-helm : Uninstalling helm] ******************************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:36
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "helm reset", "delta": "0:00:05.083432", "end": "2018-05-31 16:01:58.369972", "rc": 0, "start": "2018-05-31 16:01:53.286540", "stderr": "", "stderr_lines": [], "stdout": "Tiller (the Helm server-side component) has been uninstalled from your Kubernetes Cluster.", "stdout_lines": ["Tiller (the Helm server-side component) has been uninstalled from your Kubernetes Cluster."]}

TASK [k8s-openebs-uninstall-helm : Checking if tiller has been uninstalled successfully] *******************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:45
FAILED - RETRYING: Checking if tiller has been uninstalled successfully (10 retries left).
changed: [kubemaster01] => {"attempts": 2, "changed": true, "cmd": "kubectl get po -n \"kube-system\"", "delta": "0:00:01.054303", "end": "2018-05-31 16:03:04.569270", "rc": 0, "start": "2018-05-31 16:03:03.514967", "stderr": "", "stderr_lines": [], "stdout": "NAME                                    READY     STATUS    RESTARTS   AGE\netcd-kubemaster-01                      1/1       Running   5          1d\nkube-apiserver-kubemaster-01            1/1       Running   5          1d\nkube-controller-manager-kubemaster-01   1/1       Running   18         1d\nkube-dns-545bc4bfd4-bc67f               3/3       Running   6          1d\nkube-proxy-9fgwg                        1/1       Running   3          1d\nkube-proxy-bpg7s                        1/1       Running   26         1d\nkube-proxy-l6qtv                        1/1       Running   11         1d\nkube-proxy-rplzv                        1/1       Running   10         1d\nkube-scheduler-kubemaster-01            1/1       Running   7          1d\nkubernetes-dashboard-796487df76-2pm4m   1/1       Running   2          1d\nweave-net-p5th5                         2/2       Running   6          1d\nweave-net-p8fqh                         2/2       Running   16         1d\nweave-net-vdbw5                         2/2       Running   10         1d\nweave-net-vlw7g                         2/2       Running   4          1d", "stdout_lines": ["NAME                                    READY     STATUS    RESTARTS   AGE", "etcd-kubemaster-01                      1/1       Running   5          1d", "kube-apiserver-kubemaster-01            1/1       Running   5          1d", "kube-controller-manager-kubemaster-01   1/1       Running   18         1d", "kube-dns-545bc4bfd4-bc67f               3/3       Running   6          1d", "kube-proxy-9fgwg                        1/1       Running   3          1d", "kube-proxy-bpg7s                        1/1       Running   26         1d", "kube-proxy-l6qtv                        1/1       Running   11         1d", "kube-proxy-rplzv                        1/1       Running   10         1d", "kube-scheduler-kubemaster-01            1/1       Running   7          1d", "kubernetes-dashboard-796487df76-2pm4m   1/1       Running   2          1d", "weave-net-p5th5                         2/2       Running   6          1d", "weave-net-p8fqh                         2/2       Running   16         1d", "weave-net-vdbw5                         2/2       Running   10         1d", "weave-net-vlw7g                         2/2       Running   4          1d"]}

TASK [k8s-openebs-uninstall-helm : Deleting service account] ***********************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:54
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "kubectl delete sa tiller -n \"kube-system\"", "delta": "0:00:00.846612", "end": "2018-05-31 16:03:06.183435", "rc": 0, "start": "2018-05-31 16:03:05.336823", "stderr": "", "stderr_lines": [], "stdout": "serviceaccount \"tiller\" deleted", "stdout_lines": ["serviceaccount \"tiller\" deleted"]}

TASK [k8s-openebs-uninstall-helm : Deleting clusterolebinding] *********************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:63
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "kubectl delete clusterrolebinding tiller", "delta": "0:00:00.721009", "end": "2018-05-31 16:03:07.931464", "rc": 0, "start": "2018-05-31 16:03:07.210455", "stderr": "", "stderr_lines": [], "stdout": "clusterrolebinding \"tiller\" deleted", "stdout_lines": ["clusterrolebinding \"tiller\" deleted"]}

TASK [k8s-openebs-uninstall-helm : Get $HOME of K8s master for kubernetes user] ****************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:72
changed: [kubemaster01 -> None] => {"changed": true, "cmd": "source ~/.profile; echo $HOME", "delta": "0:00:00.016765", "end": "2018-05-31 16:03:09.007049", "rc": 0, "start": "2018-05-31 16:03:08.990284", "stderr": "", "stderr_lines": [], "stdout": "/home/ubuntu", "stdout_lines": ["/home/ubuntu"]}

TASK [k8s-openebs-uninstall-helm : Deleting the patch yaml file copied to k8s master.] *********************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:80
ok: [kubemaster01] => {"changed": false, "path": "(/home/ubuntu/update.json)", "state": "absent"}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************************************************************************************
kubemaster01               : ok=11   changed=9    unreachable=0    failed=0
```

yudaykiran commented on ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+#7) Delete the test artifacts.
+
+- name: Getting OpenEBS chart name
+  shell: helm ls --all | grep openebs | grep DEPLOYED | awk '{print $1}'
```
chartname = variable

gprasath replied:
@yudaykiran  obtaining the chart name from deployment

gprasath replied:
Obtaining the chart name based on the status and image name not the namespace name. So, it should be unique.

gprasath replied:
@yudaykiran , we fetch the chart name from the list of charts deployed and then deleting it. Do you want to specify the chart name used in 'helm installation role' here also and attempt deleting it?

gprasath commented:
Sample test run output:

Deleting the chart with name 'storage-mule'
```
giri@vagrant-host:~/openebs/e2e/ansible$ ansible-playbook cleanup-openebs.yml -vv
ansible-playbook 2.4.3.0
  config file = /home/giri/openebs/e2e/ansible/ansible.cfg
  configured module search path = [u'/home/giri/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
Using /home/giri/openebs/e2e/ansible/ansible.cfg as config file

PLAYBOOK: cleanup-openebs.yml ******************************************************************************************************************************************
1 plays in cleanup-openebs.yml

PLAY [kubernetes-kubemasters] ******************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************************
ok: [kubemaster01]
META: ran handlers

TASK [k8s-openebs-uninstall-helm : Getting OpenEBS chart name] *********************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:13
changed: [kubemaster01] => {"changed": true, "cmd": "helm ls --all | grep openebs | grep DEPLOYED | awk '{print $1}'", "delta": "0:00:01.052527", "end": "2018-06-12 16:10:41.134192", "rc": 0, "start": "2018-06-12 16:10:40.081665", "stderr": "", "stderr_lines": [], "stdout": "storage-mule", "stdout_lines": ["storage-mule"]}

TASK [k8s-openebs-uninstall-helm : Uninstall OpenEBS] ******************************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:19
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "helm del --purge \"storage-mule\"", "delta": "0:00:10.869496", "end": "2018-06-12 16:10:53.147440", "rc": 0, "start": "2018-06-12 16:10:42.277944", "stderr": "", "stderr_lines": [], "stdout": "release \"storage-mule\" deleted", "stdout_lines": ["release \"storage-mule\" deleted"]}

TASK [k8s-openebs-uninstall-helm : Checking if provisioner and api server are deleted.] ********************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:28
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "kubectl get po -n \"openebs\"", "delta": "0:00:00.819314", "end": "2018-06-12 16:10:54.849611", "rc": 0, "start": "2018-06-12 16:10:54.030297", "stderr": "", "stderr_lines": [], "stdout": "NAME                                              READY     STATUS        RESTARTS   AGE\nstorage-mule-openebs-apiserver-555c87859d-jq9w9   1/1       Terminating   1          10m", "stdout_lines": ["NAME                                              READY     STATUS        RESTARTS   AGE", "storage-mule-openebs-apiserver-555c87859d-jq9w9   1/1       Terminating   1          10m"]}

TASK [k8s-openebs-uninstall-helm : Uninstalling helm] ******************************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:37
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "helm reset", "delta": "0:00:04.508754", "end": "2018-06-12 16:10:59.998609", "rc": 0, "start": "2018-06-12 16:10:55.489855", "stderr": "", "stderr_lines": [], "stdout": "Tiller (the Helm server-side component) has been uninstalled from your Kubernetes Cluster.", "stdout_lines": ["Tiller (the Helm server-side component) has been uninstalled from your Kubernetes Cluster."]}

TASK [k8s-openebs-uninstall-helm : Checking if tiller has been uninstalled successfully] *******************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:46
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "kubectl get po -n \"kube-system\"", "delta": "0:00:00.944360", "end": "2018-06-12 16:11:02.191055", "rc": 0, "start": "2018-06-12 16:11:01.246695", "stderr": "", "stderr_lines": [], "stdout": "NAME                                    READY     STATUS     RESTARTS   AGE\netcd-kubemaster-01                      1/1       Running    8          9h\nkube-apiserver-kubemaster-01            1/1       Running    6          9h\nkube-controller-manager-kubemaster-01   1/1       Running    6          9h\nkube-dns-86f4d74b45-h9lkn               3/3       Running    8          9h\nkube-proxy-cwrkr                        1/1       Running    5          9h\nkube-proxy-s52z2                        1/1       Running    2          9h\nkube-proxy-spgtn                        1/1       NodeLost   5          9h\nkube-proxy-tj2hj                        1/1       Running    3          9h\nkube-router-4jjs5                       1/1       Running    18         9h\nkube-router-7nh4c                       1/1       Running    12         9h\nkube-router-95zhd                       1/1       Running    10         9h\nkube-router-qbzlx                       1/1       NodeLost   13         9h\nkube-scheduler-kubemaster-01            1/1       Running    3          9h\nkubernetes-dashboard-75d647ffd9-g6vpm   1/1       Running    2          9h", "stdout_lines": ["NAME                                    READY     STATUS     RESTARTS   AGE", "etcd-kubemaster-01                      1/1       Running    8          9h", "kube-apiserver-kubemaster-01            1/1       Running    6          9h", "kube-controller-manager-kubemaster-01   1/1       Running    6          9h", "kube-dns-86f4d74b45-h9lkn               3/3       Running    8          9h", "kube-proxy-cwrkr                        1/1       Running    5          9h", "kube-proxy-s52z2                        1/1       Running    2          9h", "kube-proxy-spgtn                        1/1       NodeLost   5          9h", "kube-proxy-tj2hj                        1/1       Running    3          9h", "kube-router-4jjs5                       1/1       Running    18         9h", "kube-router-7nh4c                       1/1       Running    12         9h", "kube-router-95zhd                       1/1       Running    10         9h", "kube-router-qbzlx                       1/1       NodeLost   13         9h", "kube-scheduler-kubemaster-01            1/1       Running    3          9h", "kubernetes-dashboard-75d647ffd9-g6vpm   1/1       Running    2          9h"]}

TASK [k8s-openebs-uninstall-helm : Deleting service account] ***********************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:55
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "kubectl delete sa tiller -n \"kube-system\"", "delta": "0:00:00.565807", "end": "2018-06-12 16:11:03.412438", "rc": 0, "start": "2018-06-12 16:11:02.846631", "stderr": "", "stderr_lines": [], "stdout": "serviceaccount \"tiller\" deleted", "stdout_lines": ["serviceaccount \"tiller\" deleted"]}

TASK [k8s-openebs-uninstall-helm : Deleting clusterolebinding] *********************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:64
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "kubectl delete clusterrolebinding tiller", "delta": "0:00:00.687952", "end": "2018-06-12 16:11:04.836220", "rc": 0, "start": "2018-06-12 16:11:04.148268", "stderr": "", "stderr_lines": [], "stdout": "clusterrolebinding.rbac.authorization.k8s.io \"tiller\" deleted", "stdout_lines": ["clusterrolebinding.rbac.authorization.k8s.io \"tiller\" deleted"]}

TASK [k8s-openebs-uninstall-helm : Deleting the namespace] *************************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:73
changed: [kubemaster01] => {"attempts": 1, "changed": true, "cmd": "kubectl delete ns \"openebs\"", "delta": "0:00:00.495852", "end": "2018-06-12 16:11:06.660801", "rc": 0, "start": "2018-06-12 16:11:06.164949", "stderr": "", "stderr_lines": [], "stdout": "namespace \"openebs\" deleted", "stdout_lines": ["namespace \"openebs\" deleted"]}

TASK [k8s-openebs-uninstall-helm : Get $HOME of K8s master for kubernetes user] ****************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:82
changed: [kubemaster01 -> None] => {"changed": true, "cmd": "source ~/.profile; echo $HOME", "delta": "0:00:00.005621", "end": "2018-06-12 16:11:07.290243", "rc": 0, "start": "2018-06-12 16:11:07.284622", "stderr": "", "stderr_lines": [], "stdout": "/home/ubuntu", "stdout_lines": ["/home/ubuntu"]}

TASK [k8s-openebs-uninstall-helm : Deleting the patch yaml file copied to k8s master.] *********************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:90
changed: [kubemaster01] => {"changed": true, "path": "/home/ubuntu/update.json", "state": "absent"}

TASK [k8s-openebs-uninstall-helm : Deleting the storage classes] *******************************************************************************************************
task path: /home/giri/openebs/e2e/ansible/roles/k8s-openebs-uninstall-helm/tasks/main.yaml:95
changed: [kubemaster01] => {"changed": true, "cmd": "kubectl delete -f \"https://raw.githubusercontent.com/openebs/openebs/master/k8s/openebs-storageclasses.yaml\"", "delta": "0:00:03.650474", "end": "2018-06-12 16:11:12.731532", "rc": 0, "start": "2018-06-12 16:11:09.081058", "stderr": "", "stderr_lines": [], "stdout": "storageclass.storage.k8s.io \"openebs-standalone\" deleted\nstorageclass.storage.k8s.io \"openebs-percona\" deleted\nstorageclass.storage.k8s.io \"openebs-jupyter\" deleted\nstorageclass.storage.k8s.io \"openebs-mongodb\" deleted\nstorageclass.storage.k8s.io \"openebs-cassandra\" deleted\nstorageclass.storage.k8s.io \"openebs-redis\" deleted\nstorageclass.storage.k8s.io \"openebs-kafka\" deleted\nstorageclass.storage.k8s.io \"openebs-zk\" deleted\nstorageclass.storage.k8s.io \"openebs-es-data-sc\" deleted", "stdout_lines": ["storageclass.storage.k8s.io \"openebs-standalone\" deleted", "storageclass.storage.k8s.io \"openebs-percona\" deleted", "storageclass.storage.k8s.io \"openebs-jupyter\" deleted", "storageclass.storage.k8s.io \"openebs-mongodb\" deleted", "storageclass.storage.k8s.io \"openebs-cassandra\" deleted", "storageclass.storage.k8s.io \"openebs-redis\" deleted", "storageclass.storage.k8s.io \"openebs-kafka\" deleted", "storageclass.storage.k8s.io \"openebs-zk\" deleted", "storageclass.storage.k8s.io \"openebs-es-data-sc\" deleted"]}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************************************************************************************
kubemaster01               : ok=12   changed=11   unreachable=0    failed=0
```

gprasath commented:
@ksatchit , yes it has been addressed already. This is in line with the installation task.
